### PR TITLE
Python: Port taint tests to use inline expectations

### PR DESCRIPTION
**Pull request by RasmusWL**
The meat of this PR is described in the new python/ql/test/experimental/meta/InlineTaintTest.qll file:

> Defines a InlineExpectationsTest for checking whether any arguments in
> `ensure_tainted` and `ensure_not_tainted` calls are tainted.
>
> Also defines query predicates to ensure that:
> - if any arguments to `ensure_not_tainted` are tainted, their annotation is marked with `SPURIOUS`.
> - if any arguments to `ensure_tainted` are not tainted, their annotation is marked with `MISSING`.
>
> The functionality of this module is tested in `ql/test/experimental/meta/inline-taint-test-demo`.

One of the things I liked the best about the old `TestTaint.expected` files was that it was easy to see how many failures we had. However, it's annoying to review code that only changes the `.expected` file, since you need to manually look up the `.py` file to get the full picture of what is now handled (see [here](https://github.com/github/codeql/pull/5681/files#diff-17d8ae287b09c0f1ccd0ebbcabdc958926d149161e4d206f41c83a652b602d29) as an example).

To ensure we can still easily get an overview of failures, I introduced the requirement to add `SPURIOUS` and `MISSING` annotations. So it should be easy to grep for `(SPRURIOUS|MISSING): tainted` and find all cases our taint-steps could be improved.

In regards to reviewing, it will be a pain to go through all files individually to see that everything is the same as before (I know, since I did this myself :cry:) -- so I'll suggest that you look at the demo, and if you are convinced that the new `InlineTaintTest.qll` works properly, this PR should be good to go if tests pass :crossed_fingers:

### CODE REVIEW

File: python/ql/test/experimental/meta/inline-taint-test-demo/taint_test.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    ensure_tainted(
+        should_be_tainted,
+    )
+
+    # simulating handling something we _want_ to treat at untainted, but we currently treat as tainted
+    should_not_be_tainted = "pretend this is now safe" + ts
+
+    # This annotation _should_ have used `SPURIOUS`, which will be alerted on in the .expected output
+    ensure_not_tainted(
+        should_not_be_tainted # $ tainted
```
Comment (yoff) [inline python/ql/test/experimental/meta/inline-taint-test-demo/taint_test.py:36]: A better test might be if we forgot to write any annotation at all.

File: python/ql/test/library-tests/frameworks/tornado/taint_test.py [hunk 2]
```diff
@@ -3,73 +3,73 @@
 
 class TaintTest(tornado.web.RequestHandler):
     def get(self, name = "World!", number="0", foo="foo"):  # $ requestHandler routedParameter=name routedParameter=number
-        ensure_tainted(name, number)
+        ensure_tainted(name, number) # $ tainted
```
Comment (yoff) [inline python/ql/test/library-tests/frameworks/tornado/taint_test.py:6]: Here you now have to split into two lines.

**Review by yoff — CHANGES_REQUESTED** [review]
This is a great initiative 💪 
In the longer run, I would prefer if we could generalise the interface to `InlineExpectationTest` so that we could recognise `ensure_tainted` and `ensure_not_tainted` as annotations directly. But given how you have coded around this very nicely (with your extra consistency predicates), it is more a nice-to-have at this point.

Only two minor points, which came piecemeal because I pressed the wrong button..

File: python/ql/test/experimental/meta/inline-taint-test-demo/taint_test.py (hunk 1 above)
Comment (RasmusWL) [reply python/ql/test/experimental/meta/inline-taint-test-demo/taint_test.py:36]: I don't agree. With this setup, only things that are (or should be) tainted should have any annotations.

File: python/ql/test/library-tests/frameworks/tornado/taint_test.py (hunk 2 above)
Comment (RasmusWL) [reply python/ql/test/library-tests/frameworks/tornado/taint_test.py:6]: no, this just means that all arguments on that line are tainted :wink:

File: python/ql/test/experimental/meta/inline-taint-test-demo/taint_test.py (hunk 1 above)
Comment (yoff) [reply python/ql/test/experimental/meta/inline-taint-test-demo/taint_test.py:48]: Ah, right, arguments to `ensure_not_tainted` which are indeed not tainted, do not get any annotation. That makes perfect sense.

**Review by yoff — COMMENTED** [review]
LGTM, happy to merge once the conflicts are resolved.

**Comment by RasmusWL** [comment]
Interesting part of merge is https://github.com/github/codeql/pull/5687/commits/63a2657aef971ae082c3b8b50b40de9db4284143#diff-b952effafd539a00838363ef1cc0369a465f5f3687d60c839ceb19d3cf970da6

**Review by yoff — APPROVED** [review]
LGTM

**Comment by yoff** [comment]
> Interesting part of merge is [63a2657#diff-b952effafd539a00838363ef1cc0369a465f5f3687d60c839ceb19d3cf970da6](https://github.com/github/codeql/commit/63a2657aef971ae082c3b8b50b40de9db4284143#diff-b952effafd539a00838363ef1cc0369a465f5f3687d60c839ceb19d3cf970da6)

Thanks :sweat_smile:

**Comment by RasmusWL** [comment]
:shipit:

**Pull request merged**